File: src/algorithms/stacks-queues/validation/min-remove-to-make-valid/sources/min-remove-to-make-valid.ts
// Min Remove to Make Valid — use a stack of indices to track unmatched '(' and a set for unmatched ')'
function minRemoveToMakeValid(inputString: string): string {
  const unmatchedOpenIndices: number[] = []; // @step:initialize
  const unmatchedCloseIndices: Set<number> = new Set(); // @step:initialize
  for (let charIdx = 0; charIdx < inputString.length; charIdx++) {
    const char = inputString[charIdx]!; // @step:visit
    if (char === "(") {
      unmatchedOpenIndices.push(charIdx); // @step:push
    } else if (char === ")") {
      if (unmatchedOpenIndices.length > 0) {
        unmatchedOpenIndices.pop(); // @step:pop
      } else {
        unmatchedCloseIndices.add(charIdx); // @step:mismatch
      }
    }
  }
  // Remaining indices in the stack are unmatched opening brackets
  const unmatchedIndices: Set<number> = new Set([
    ...unmatchedOpenIndices,
    ...unmatchedCloseIndices,
  ]); // @step:mismatch
  let result = ""; // @step:complete
  for (let charIdx = 0; charIdx < inputString.length; charIdx++) {
    if (!unmatchedIndices.has(charIdx)) {
      result += inputString[charIdx]; // @step:complete
    }
  }
  return result; // @step:complete
}

File: src/algorithms/stacks-queues/validation/min-remove-to-make-valid/sources/min-remove-to-make-valid.go
// Min Remove to Make Valid — use a stack of indices to track unmatched '(' and a set for unmatched ')'
package main

import "fmt"

func minRemoveToMakeValid(inputString string) string {
	unmatchedOpenIndices := []int{} // @step:initialize
	unmatchedCloseIndices := map[int]bool{} // @step:initialize
	runes := []rune(inputString)
	for charIdx := 0; charIdx < len(runes); charIdx++ {
		ch := runes[charIdx] // @step:visit
		if ch == '(' {
			unmatchedOpenIndices = append(unmatchedOpenIndices, charIdx) // @step:push
		} else if ch == ')' {
			if len(unmatchedOpenIndices) > 0 {
				unmatchedOpenIndices = unmatchedOpenIndices[:len(unmatchedOpenIndices)-1] // @step:pop
			} else {
				unmatchedCloseIndices[charIdx] = true // @step:mismatch
			}
		}
	}
	// Remaining indices in the stack are unmatched opening brackets
	unmatchedIndices := map[int]bool{} // @step:mismatch
	for idx, val := range unmatchedCloseIndices {
		unmatchedIndices[idx] = val
	}
	for _, idx := range unmatchedOpenIndices {
		unmatchedIndices[idx] = true
	}
	result := "" // @step:complete
	for charIdx := 0; charIdx < len(runes); charIdx++ {
		if !unmatchedIndices[charIdx] {
			result += string(runes[charIdx]) // @step:complete
		}
	}
	return result // @step:complete
}

func main() {
	fmt.Println(minRemoveToMakeValid("lee(t(c)o)de)"))
}
